sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2DataFrameSuite.scala

```diff
@@ -1260,26 +1251,81 @@ class DataSourceV2DataFrameSuite
         Row(3, "c", 30, null),
         Row(4, "d", 40, "x")))
 
-      // join between df1 and df3 should fail as refreshing versions is not
```

Comment:
This is arguably the most questionable part in DSv1 behavior today. This is allowed in Delta, for instance. I am going back and forth. For now, I decided to follow what we had in Spark before as it can be justified that Dataset pins / remembers its schema after analysis and it can never change.

Reply:
Note that we still refresh to use consistent versions and prohibit removing columns or changing nullability.